nmigen/build/generic_platform.py

```diff
@@ -0,0 +1,99 @@
+__all__ = ["Pins", "Subsignal", "DiffPairs", "Resource", "Connector"]
+
+
+class Pins:
+    def __init__(self, *identifiers, dir="io"):
+        for i in identifiers:
+            if not isinstance(i, str):
+                raise TypeError("Identifier must be a string, not {!r}".format(i))
+        self.identifiers = " ".join(identifiers).split(" ")
+
+        if dir not in ("i", "o", "io"):
+            raise TypeError("Direction must be one of \"i\", \"o\" or \"io\", not {!r}"
+                            .format(dir))
+        self.dir = dir
+
+    def __len__(self):
+        return len(self.identifiers)
+
+    def __repr__(self):
+        return "(pins {} \"{}\")".format(" ".join(self.identifiers), self.dir)
+
+
+class DiffPairs:
+    def __init__(self, *pairs, dir="io"):
+        p, n = zip(*(pair.split(" ") for pair in pairs))
+        self.p = Pins(*p, dir=dir)
+        self.n = Pins(*n, dir=dir)
+
+        self.dir = self.p.dir
+
+    def __len__(self):
+        return len(self.p)
+
+    def __repr__(self):
+        return "(diffpairs {} {})".format(self.p, self.n)
+
+
+class Subsignal:
+    def __init__(self, name, *io, extras=()):
+        self.name = name
+
+        self.has_pins       = len(io) == 1 and isinstance(io[0], Pins)
+        self.has_diffpairs  = len(io) == 1 and isinstance(io[0], DiffPairs)
+        self.has_subsignals = io and all(isinstance(c, Subsignal) for c in io)
+
+        if not (self.has_pins or self.has_diffpairs or self.has_subsignals):
+            raise TypeError("I/O constraints must be either Pins or DiffPairs or "
+                            "multiple Subsignal, not {!r}"
+                            .format(io))
+        self.io = io
+
+        for c in extras:
+            if not isinstance(c, str):
+                raise TypeError("Extra contraint must be a string, not {!r}".format(c))
+        self.extras = list(extras)
+
+        if self.has_subsignals:
+            for sub in self.io:
+                sub.extras += self.extras
+
+    def __repr__(self):
+        return "(subsignal {} {} {})".format(self.name, " ".join(map(repr, self.io)), self.extras)
+
+
+class Resource(Subsignal):
```

Comment:
Resources are just "top-level" Subsignals. 

In case multiple IO resources share the same name, they have an additional `number` attribute to identify them (e.g. `platform.request("user_led", 1)`).